### PR TITLE
docs(ci): Deploy Previews

### LINKED ISSUE #1871: [FR] Documentation PR preview deploys
# Feature Request

There was interest in having PRs for documentation updates to have CI add a comment with link to preview the changes visually.

This has been discussed during the Github Wiki to MkDocs (via Github Pages) migration work. A free service can provide this functionality, and be integrated via Github Actions for CI.

## Context

I have offered to contribute this, however as this project now exists as an organization, **it requires a bit more effort on behalf of organization members to make arrangements for being granted an open-source organization team plan (free if approved)** on either of the two popular services:

- Netlify is more open about their support for this [here](https://www.netlify.com/legal/open-source-policy), with requirements and actions required to get approval.
- Vercel is my original choice, they briefly mention some form of support for open-source projects beyond the personal hobby account [near the bottom of their pricing page](https://vercel.com/pricing).

I've done most of my research into supporting Vercel, but from what I can tell it should be a similar process with Netlify:

- Connecting the organization/team and project to the service.
- A configuration file to disable default default branch and PR behaviour (lacks context of when it should be triggered)
- A third-party Github Action that only triggers under the PR event for PRs that actually touch the docs.

### Is your Feature Request related to a Problem?

It would assist with PR reviews related to documentation contributions.

### Describe the Solution you'd like

Leveraging the free tier support of Vercel or Netlify services with the preview deploy links.

The organization has already expressed interest in keeping deployment of docs tied to Github Pages. That does have some slight disadvantages compared to these services feature wise (eg rewrites, redirects, 404 handling, cache control, etc), but with the above plan of action the service can specifically only deploy previews AFAIK.

### Are you going to implement it?

Yes, because I know the probability of someone else doing it is low and I can learn from it.

### What are you going to contribute??

Everything required to enable the feature.

## Additional context

~~I cannot proceed further with either service until the organization~~ (**EDIT:** I've been made a member, I'll take care of it) decides on a service to use and has taken action to be granted approval for the open-source free team plan.

### Alternatives you've considered

I'm not sure of what other alternatives we have, other than more DIY, but this would raise the difficulty and add more maintenance burden due to low bus factor than using one of these popular services would.

### Who will that Feature be useful to?

Maintainers and/or reviewers of pull requests for documentation.

### What have you done already?

Investigated required tasks and effort to contribute the feature.

### PR DESCRIPTION
**Pull request by polarathene**
# Description

**Feature:** When a PR contributes to our docs, a preview link will be added as a comment via Github Actions.

---

Each PR will have a unique prefix alias of `pullrequest-N--` (_where N is the PR number_) to our docs preview site at [dms-doc-previews.netlify.app](https://dms-doc-previews.netlify.app/) (_404 page is intentional, only the preview aliases for PRs will have actual content_).

Example: `https://pullrequest-42--dms-doc-previews.netlify.app`

---

Details of interested are covered below. The workflows have some inlined documentation via comments, and each commit should be fairly clean with useful commit messages.

## Sponsored plan + requirements to retain

<details>
<summary>Netlify has sponsored the docker-mailserver organization with an OSS team account.</summary>

To keep this sponsorship, we must comply with the [plan policy](https://www.netlify.com/legal/open-source-policy); we qualified for sponsorship so currently meet the requirements. The Netlify branding mention is injected into build config for PRs only and can be found in the footer section on all pages.

We continue to deploy the production versions of documentation to Github Pages, but will utilize their service for deploying preview builds of PRs touching the documentation.

</details>

## Billing / Plan limits

<details>
<summary>There are two billing concerns that still apply if exceeded</summary>

- [Bandwidth usage](https://docs.netlify.com/accounts-and-billing/billing/#bandwidth-usage) (400GB monthly)
  - This is regarding bandwidth transferring data to viewers. Our site is under 2MB in total, but this is considerably less over the wire as Netlify compresses responses/assets with gzip and brotli where appropriate. Caching is also leveraged on the client side.
  - As it's only being used for previewing PR contributions for docs, this is expected to be very low and a non-concern.
- [Build Minutes](https://docs.netlify.com/accounts-and-billing/billing/#builds-usage) (1000 minutes per month)
  - This shouldn't be a concern. Our builds don't take long, but it could more easily be abused maliciously or through very high frequency of pushing commits.
  - Github Actions is handling builds instead, and then transferring the build to Netlify via their deploy API. This bandwidth is not part of the billing limit, and does not impact the build time.
  - To further guard against high frequency commits to a PR, Github Actions `concurrency` feature is in use to cancel any existing instance of the workflow for the same PR.

In the rare event one of those limits are exceeded to incur charges, see [overdue accounts](https://docs.netlify.com/accounts-and-billing/billing/#overdue-accounts). If unpaid for 2 weeks, the account is restricted (no new deploys can be performed), within 4 weeks the account is suspended and the hosted deploys will no longer be available.

</details>

## Github Actions - Workflow Security

This has been documented into the workflow to make the concern clear to any maintainers. Github authored a [blog post](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/) recently about workflows with PRs.

<details>
<summary>Multi-staged workflow for security complicates the workflow a tad</summary>

`pull_request` event runs with secrets if the PR author is a trusted collaborator on the project, but for third-party community contributions untrusted code may be a concern for stealing secrets, thus Github restricts access.

Instead the `pull_request` workflow should handle running any untrusted code such as performing a build, and using a secure workflow (that's consistent with secrets access regardless of contributor) for handling anything requiring secrets that is not at the risk of running untrusted code from a contributor.

`workflow_run` is used as a 2nd stage of the PR workflow for this. It does complicate the process a bit more with additional steps, but is required for us to securely support deploy previews with all documentation PRs. This workflow will always run from the default branch only, and this affects some CI context such as the latest commit SHA.

Some actions have features that aren't compatible with this approach (eg `nwtgck/actions-netlify`). Either the maintainer will resolve them or I may contribute a fix. In the meantime that incompatibility has been minimized via additional actions in the workflow, while others like the deployment environment commit being logged incorrectly is a minor inconvenience.

</details>

## Misc

<details>
<summary>These previews will remain reachable long after the PRs are merged</summary>

- There is no automated cleanup, but it should not present any issue AFAIK.
- As these are considered deploy previews by Netlify and not production marked builds, Netlify response headers advise search engines to not index these deployments.

---

</details>

<details>
<summary>Some small related improvements were made</summary>

- Removed unnecessary build output and shared this logic in a shell script for running the build between preview and production deploy workflows. Version bumps to the docker image will be centralized there.
- Cleaned up source logo files and optimized production versions.
- Added a custom 404 page.
  - Not really needed, this was added to the Netlify project to initialize a site instead of connecting to the repo.
  - Figured it might as well be used as the docs 404 page at that point, it should not cause any maintenance burden AFAIK, but if source is required let me know.

</details>

Fixes #1871

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by polarathene** [comment]
[Two secrets need to be added to this project](https://github.com/nwtgck/actions-netlify#required-inputs-and-env) for deployment to be successful, I have invited two maintainers and notified them of this requirement. Once that is done this PR is ready for merging.

**Comment by aendeavor** [comment]
LGTM 👍 

@wernerfred could you have a look at the secrets?

**Comment by wernerfred** [comment]
Insane, you gave me much insight and inspirations on how to tackle problems i had with triggers based on different worklfows. Thanks also for the linked blog post!
Is anything missing right now?

**Review by wernerfred — APPROVED** [review]
Thanks for this contribution!
I could follow the changes/logic in general and it LvGTM 👍🏻 

Could be that there are some minors that occur only during workflow run but should be easy fixable then.

**Review by aendeavor — APPROVED** [review]

File: .github/workflows/scripts/docs/update-versions-json.sh
Comment (casperklein) [inline .github/workflows/scripts/docs/update-versions-json.sh:6]: Small styling thing: this should be `function _update_versions_json`

**Comment by polarathene** [comment]
> Insane, you gave me much insight and inspirations on how to tackle problems i had with triggers based on different worklfows.

You're welcome! It took me a while to figure it all out for non-personal usage :smile: 

> Is anything missing right now?

Not that I know of. It went through multiple revisions over the past two months when I had time for it. I have not created a 2nd account to test a PR from a non-collaborator, but it should work or be addressable with a follow up PR if needed.

Deployment should otherwise work well, which has been tested and verified on my test repo.

File: .github/workflows/scripts/docs/update-versions-json.sh
Comment (polarathene) [inline .github/workflows/scripts/docs/update-versions-json.sh:6]: Lazily applying change:

```suggestion
function _update_versions_json
```

**Review by wernerfred — APPROVED** [review]

**Review by aendeavor — APPROVED** [review]

**Comment by wernerfred** [comment]
Shouldn't have this PR already build a preview? The Build Preview workflow has passed

**Comment by polarathene** [comment]
> But could also be that only the file that was previously there (you only renamed it) was triggered and not the new one.

If you rename a workflow, at least with the `name` field IIRC it will update the name to that. It did seem to run [the new PR workflow steps](https://github.com/docker-mailserver/docker-mailserver/runs/2628799261?check_suite_focus=true) however.

The 2nd stage which takes the build and deploys it via `workflow_run` is run from the default branch, presumably that was not available at this point.

> We'll see with the LDAP docs PR ;)

:wink: :pray:

**Comment by wernerfred** [comment]
> 😉 🙏

Insanity! 🎉  https://github.com/docker-mailserver/docker-mailserver/pull/1921#issuecomment-844981421

**Pull request merged**
